Added instructions for reviewer - OOP for data science (#204)
Umuzi-org/ACN-syllabus: content/projects/data-science-specific/oop-for-data-science/_index.md

5. [OPTIONAL] Create a class called `InfluenceCalculator` that takes a model fit object as an argument and has methods for calculating and displaying leverage values and Cook's distance. Call these methods `leverage`, `cooks_distance`. This is easiest to do using the package `statsmethods` to create the linear regression model object. Are there any observations that may be influencing the models?

_Tip:_ These classes should be able to work with any regression problem. You can re-use these classes in upcoming assignments when inspecting models and you can also make classes in future that can work for both regression and classification problems.

## Instructions for reviewer

> 1. One of the main purposes of OOP is the reusableness of code.  Therefore it is important that learners do not

Reviewer: reusableness should probably be reusability

Reviewer 2: I agree

   create new classes, objects or code when the same result could have been achieved if the learner understood OOP.
   Since Inheritance and Polymorphism are core concepts of OOP, adequate understanding of these concepts will be revealed when
   learners make use of creating classes which inherit from a parent class, just like in Animals part 1 where both Cat and Dog

>    were classes which inherited from class Animal. The same then is true for Polymorphism, an easy example is where the sound()

Reviewer: Referencing other projects could create technical debt, say we remove the animals project from  the syllabus, then this would also need to be updated

Reviewer 2: If Animals part 1 is a precursor to this project then it should be listed as a prerequisite

_Tip:_ These classes should be able to work with any regression problem. You can re-use these classes in upcoming assignments when inspecting models and you can also make classes in future that can work for both regression and classification problems.

## Instructions for reviewer

1. Learners should easily apply the skills they learned in Animals part 1 and Animals part 2 with reference to OOP.
   It should not be the case that a learner is unable to do the basics, for instance, creating a class and having
   a 'constructor'/__init__() function in order to initialize and object when it is created from the class.

2. One of the main purposes of OOP is the reusableness of code.  Therefore it is important that learners do not
   create new classes, objects or code when the same result could have been achieved if the learner understood OOP
   and therefore made use of what was already developed and available.

> 3. Learners should show a clear understanding of Inheritance from parent classes.  This would be most evidently
>    showed by the use of super() and the way in which a child class is created, seeing that super() is not the only way
>    of going about it, a learner could for instance have "class child_class(parent_class):".

Reviewer 3: Remember that learners will be coming here to read the instructions, so we don't want to give away too much in the instructions for reviews. The second part of this is basically telling them something that they should be figuring out.
```suggestion
3. Learners should show a clear understanding of Inheritance from parent classes. 
```

4. The same rules apply with regards to plots as with previous projects.  Plots should be neat and self-explanatory,
   plots should have clear and concise labeling and the data represented should be easy to read and intuitive.  There
   should not be a histogram plot when a scatter plot is needed and vice-versa.

5. Learners should show an adequate understanding of decorators.  Although decorators can be tricky, it might be that
   various functions and functionality overlap, these slight changes in functionality can then easily be catered for
   by the appropriate use of decorators.  This is then another way to ensure that learners don't create unecessary functions,

>    hence, the code stays DRY.

Reviewer 3: This project can be done without decorators, and since we don't explicitly mention that they should use then we shouldn't require that they show they know how to use them.

6. Since Inheritance and Polymorphism are core concepts of OOP, adequate understanding of these concepts will be revealed when
   learners make use of creating classes which inherent from a parent class, just like in Animals part 1 where both Cat and Dog
   were classes which inherited from class Animal. The same then is true for Polymorphism, an easy example is where the sound()
   function in Animals part 1. had different execution based on if the object which called the sound() function was a Cat or a Dog

>    class.

Reviewer 3: I think this can replace points 2 and 3 above. It clearly states the concepts their meant to show, and uses examples from a pre-requisite project which is great. That way we are not giving too many hints on how apply these concepts for this project in the instructions.

5. [OPTIONAL] Create a class called `InfluenceCalculator` that takes a model fit object as an argument and has methods for calculating and displaying leverage values and Cook's distance. Call these methods `leverage`, `cooks_distance`. This is easiest to do using the package `statsmodels` to create the linear regression model object. Are there any observations that may be influencing the models?

_Tip:_ These classes should be able to work with any regression problem. You can re-use these classes in upcoming assignments when inspecting models and you can also make classes in future that can work for both regression and classification problems.

## Instructions for reviewer

1. One of the main purposes of OOP is the reusability of code.  Therefore it is important that learners do not
   create new classes, objects or code when the same result could have been achieved if the learner understood OOP.
   
2. Learners should show a clear understanding of Inheritance from parent classes.  This would be most evidently
   showed by the use of super() and the way in which a child class is created.

> 3. Learners should show an adequate understanding of decorators if they make use of them, although decorators can be tricky, 
>    The use of decorators would then be another way to ensure that learners don't create unnecessary functions,
>    hence, the code stays DRY.  Note, the use of decorators is Not a pre-requisite for this project.

Reviewer 3: ```suggestion
3. Decorators are another way to ensure that learners don't create unnecessary functions, so the code stays DRY.  
   If the learner chooses to use decorators then they should use decorators properly. Note, the use of decorators is not required for this project.
```

_Tip:_ These classes should be able to work with any regression problem. You can re-use these classes in upcoming assignments when inspecting models and you can also make classes in future that can work for both regression and classification problems.

## Instructions for reviewer

1. One of the main purposes of OOP is the reusability of code.  Therefore it is important that learners do not
   create new classes, objects or code when the same result could have been achieved if the learner understood OOP.
   Since Inheritance and Polymorphism are core concepts of OOP, adequate understanding of these concepts will be revealed when
   learners make use of creating classes which inherit from a parent class.  The same then is true for Polymorphism.
   
2. Learners should show a clear understanding of Inheritance from parent classes.  This would be most evidently
   showed by the use of super() and the way in which a child class is created, seeing that super() is not the only way

>    of going about it.

Reviewer 2: Point number 2 is a bit challenging to understand. Are you saying that the reviewer should make sure that super is used? Or are you saying that the learner should show that they dont need to use super?

> 3. The same rule applies with regards to plots as with previous projects.  Plots should be neat and self-explanatory,

Reviewer 2: What same rule? We were talking about inheritance until now. Do plots inherit too?  

Remember that this needs to set the reviewer up for success.

   plots should have clear and concise labeling and the data represented should be easy to read and intuitive.  There
   should not be a histogram plot when a scatter plot is needed and vice-versa.

> 4. Learners should show an adequate understanding of decorators if they make use of them, although decorators can be tricky, 

Reviewer 2: What should the reviewer do here? Would it be safe to paraphrase this as:

If the learner chooses to use decorators then they should use decoraters properly. Use of decorators is not required.
